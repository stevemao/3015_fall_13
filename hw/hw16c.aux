\relax 
\citation{knuthv1}
\bibstyle{plain}
\bibdata{./homework}
\bibcite{knuthv1}{1}
