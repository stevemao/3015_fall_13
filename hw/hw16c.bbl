\begin{thebibliography}{1}

\bibitem{knuthv1}
Donald Knuth.
\newblock {\em The Art of Computer Programming: Fundamental Algorithms},
  volume~1.
\newblock Addison Wesley, Reading, MA, 1973.

\end{thebibliography}
