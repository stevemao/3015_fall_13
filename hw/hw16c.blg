This is BibTeX, Version 0.99c (Web2C 7.5.4)
The top-level auxiliary file: hw16c.aux
The style file: plain.bst
Database file #1: ./homework.bib
You've used 1 entry,
            2118 wiz_defined-function locations,
            503 strings with 4096 characters,
and the built_in function-call counts, 252 in all, are:
= -- 21
> -- 7
< -- 1
+ -- 4
- -- 2
* -- 14
:= -- 49
add.period$ -- 3
call.type$ -- 1
change.case$ -- 3
chr.to.int$ -- 0
cite$ -- 1
duplicate$ -- 12
empty$ -- 23
format.name$ -- 2
if$ -- 50
int.to.chr$ -- 0
int.to.str$ -- 1
missing$ -- 2
newline$ -- 8
num.names$ -- 2
pop$ -- 6
preamble$ -- 1
purify$ -- 3
quote$ -- 0
skip$ -- 7
stack$ -- 0
substring$ -- 6
swap$ -- 2
text.length$ -- 1
text.prefix$ -- 0
top$ -- 0
type$ -- 2
warning$ -- 0
while$ -- 2
width$ -- 2
write$ -- 14
