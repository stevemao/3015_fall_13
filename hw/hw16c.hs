
data Rose a = Node a [Rose a] deriving (Eq,Show,Ord)
data Btree a = Empty | Fork a (Btree a) (Btree a) deriving (Eq,Show,Ord)

nat :: Rose a -> Btree a
nat (Node x []) = Empty              -- replace "Empty" with your code
nat (Node x (xt:xts)) = Empty        -- replace "Empty" with your code

inv :: Btree a -> Rose a
inv (Fork z Empty Empty) = Node 0 [] -- replace "Node 0 []" with you own code
inv (Fork z l r) = Node 0 []         -- replace "Node 0 []" with you own code



preorder_rose (Node x xts) = x : (concat (map preorder_rose xts))

preorder_Btree Empty = []
preorder_Btree (Fork x xt yt) = x : preorder_Btree xt ++ preorder_Btree yt


t1 = Node 1 []
t2 = Node 1 [Node 2 []]
t3 = Node  1 [Node 2 [Node 3 []]]
t4 = Node  1 [Node 2 [Node 3 []],Node 4 []]
t5 = Node  1 [Node 2 [Node 3 []],Node 4 [Node 5 []]]
t6 = Node 1 [t1,t2,t3,t4,t5]
t7 = Node 1 [Node 2 [], Node 3 [], Node 4 []]

trees = [t1,t2,t3, t4,t5,t6,t7]

-- if the functions are inverses then, for any rose tree t the
-- following should hold: (inv . nat) t = t and also, the preorder
-- traversals of the rose tree t should be the same as the preorder
-- traversal of the transformed btree.  You can test your code by
-- mapping the following function over the list trees.


test x = (inv . nat) x == x  && preorder_rose x == preorder_Btree (nat x)



