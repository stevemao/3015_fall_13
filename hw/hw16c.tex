\documentclass[11pt]{article}
\usepackage{qtree}
\include{header}


\begin{document}
\homework{16}{2 November}

This is an exercise in some rather complex recursion on Rose trees and Binary
trees. You will implement the mapping (similar to the one described in class)
of Rose trees to Binary Trees and the inverse mapping binary trees to rose
trees.

\begin{program*}
\> data Rose a = Node a [Rose a] deriving (Eq,Show) \\
\> data BTree a = Empty | Fork a (BTree a) (BTree a) deriving (Eq,Show) 
\end{program*}

The mapping is called the {\em{natural correspondence}} between trees and
binary trees \cite[pp.332]{knuthv1}.  Given a rose tree (with arbitrary but
finite branching) you construct a binary tree where the left fork for each node
in the binary tree is the translation of the leftmost child of the
corresponding node in the rose tree and the right fork is the binary tree
resulting form translating the right sibling of the node in the translated rose
tree.  Since root nodes in rose trees have no siblings, the binary trees
resulting from the translation always have the form {\tt{(Fork x t Empty)}}.

Here are some examples:\ \\

\begin{example}\ \\
\begin{program*}
\>         (Rose 1 [Rose 2 []]) $\stackrel{nat}{\longrightarrow}$  Fork 1 (Fork 2 Empty Empty) Empty \\ 
\end{program*}
\begin{centering}
\begin{tabular}{ccc}
\hspace{1.5in}\Tree [.1 [ 2 ] ] & {\mbox{{\hspace{1.25in}}}}&  \Tree [.1 [ [.2 [ $\bot$ ] [ $\bot$ ] ] ] [ $\bot$ ]  ] \\
\end{tabular}
\end{centering}
\end{example}

\begin{example}\ \\
\begin{program*}
\> (Node 1 [Node 2 [Node 3 []]]) $\stackrel{nat}{\longrightarrow}$  Fork 1 (Fork 2 (Fork 3 Empty Empty) Empty) Empty \\
\end{program*}
\begin{tabular}{ccc}
\hspace{1.5in}\Tree [.1 [.2 [ 3 ] ] ] & {\mbox{{\hspace{1.25in}}}}  &  \Tree [.1 [ [.2  [.3  [ $\bot$ ]  [ $\bot$ ] ] [ $\bot$ ] ] ] [ $\bot$ ]  ] \\
\end{tabular}
\end{example}
\newpage

\begin{example}\ \\
\begin{program*}
\> (Node 1 [Node 2 [Node 3 []],Node 4 []]) $\stackrel{nat}{\longrightarrow}$ \\
\>     Fork 1 (Fork 2 (Fork 3 Empty Empty) (Fork 4 Empty Empty)) Empty
\end{program*}
\begin{tabular}{ccc}
\hspace{1.25in}\Tree [.1 [ [.2 [ 3 ] ] ] [ 4 ]  ] 
& {\mbox{\hspace{1.5in}}} &  
\Tree [.1  [ [.2 [ [.3 [ $\bot$ ] [ $\bot$ ] ] ] [ [.4  [ $\bot$ ]  [ $\bot$ ] ] ] ] ] [ $\bot$ ] ] \\
\end{tabular}
\end{example}

\begin{example}\ \\
\begin{program*}
\> Rose 1 [Rose 2 [],Rose 3 [],Rose 4 []] $\stackrel{nat}{\longrightarrow}$ \\
\>     Fork 1 (Fork 2 Empty (Fork 3 Empty (Fork 4 Empty Empty))) Empty
\end{program*}
\begin{tabular}{ccc}
\hspace{1.25in}\Tree [.1 [ [ 2 ] [ 3 ]  [ 4 ]  ] ]
& {\mbox{\hspace{1.5in}}} &  
\Tree [.1  [ [.2  [ $\bot$ ] [.3  [ $\bot$ ] [.4 [ $\bot$ ]  [ $\bot$ ] ] ] ] [ $\bot$ ] ] ]
\end{tabular}

\end{example}

\exercise{Write a function {\tt{nat::Rose a $\rightarrow$ Btree a}} that
implements the natural transformation by recursion on the structure of rose
trees.}


\exercise{Write a function {\tt{inv::Btree a $\rightarrow$ Rose a }} that implements the inverse mapping.}

\bibliographystyle{plain}
\bibliography{./homework}

\end{document}


% Local Variables:
% mode:latex
% comment-column:0
% comment-start: "\\>  "
% comment-end: "\\\\"
% compile-command: "pdflatex hw16c"
% fill-column:79
% End:




