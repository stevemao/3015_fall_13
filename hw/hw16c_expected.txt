Prelude> :l hw16.hs
[1 of 1] Compiling Main             ( hw16c.hs, interpreted )
Ok, modules loaded: Main.
*Main> map test trees
[True,True,True,True,True,True,True]
*Main> :t nat
nat :: Rose t -> Btree t
*Main> :t inv
inv :: Btree t -> Rose t
*Main> map nat trees
map nat trees
[Fork 1 Empty Empty,Fork 1 (Fork 2 Empty Empty) Empty,Fork 1 (Fork 2 (Fork 3 Empty Empty) Empty) Empty,Fork 1 (Fork 2 (Fork 3 Empty Empty) (Fork 4 Empty Empty)) Empty,Fork 1 (Fork 2 (Fork 3 Empty Empty) (Fork 4 (Fork 5 Empty Empty) Empty)) Empty,Fork 1 (Fork 1 Empty (Fork 1 (Fork 2 Empty Empty) (Fork 1 (Fork 2 (Fork 3 Empty Empty) Empty) (Fork 1 (Fork 2 (Fork 3 Empty Empty) (Fork 4 Empty Empty)) (Fork 1 (Fork 2 (Fork 3 Empty Empty) (Fork 4 (Fork 5 Empty Empty) Empty)) Empty))))) Empty,Fork 1 (Fork 2 Empty (Fork 3 Empty (Fork 4 Empty Empty))) Empty]
*Main> 